<launch>
    <group ns="attract">
      <node name="attract_nodelets" pkg="nodelet" type="nodelet"
        args="manager"
      />

      <node name="attract" pkg="nodelet" type="nodelet"
        args="load hri_behaviors/Attract attract_nodelets">
        <remap from="pose_to_attract" to="pose"/>
      </node>

      <node name="pose_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor attract_nodelets 
          /sensors/kinect/user_face_pose pose"/>

      <node name="cmd_vel_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor attract_nodelets 
          cmd_vel cmd_vel_out"/>
      <param name="cmd_vel/abtr_priority" value="12"/>
      <node name="do_gesture_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor attract_nodelets 
          do_gesture do_gesture_out"/>
      <param name="do_gesture/abtr_priority" value="12"/>
      <node name="voice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor attract_nodelets 
          voice voice_out"/>
      <param name="voice/abtr_priority" value="12"/>
      <node name="emotion_choice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor attract_nodelets 
          emotion_choice emotion_choice_out"/>
      <param name="emotion_choice/abtr_priority" value="12"/>
      <node name="pan_angle_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor attract_nodelets 
          pan_angle pan_angle_out"/>
      <param name="pan_angle/abtr_priority" value="12"/>
    </group>

    <group ns="check_person">
      <node name="check_person_nodelets" pkg="nodelet" type="nodelet"
        args="manager"
      />

      <node name="check_person" pkg="nodelet" type="nodelet"
        args="load hri_behaviors/CheckPerson check_person_nodelets">
        <remap from="pose_to_check" to="pose"/>
        <remap from="human_pose" to="/sensors/kinect/user_face_pose"/>
      </node>

      <node name="pose_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor check_person_nodelets 
          pose_to_check pose"/>

      <node name="cmd_vel_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor check_person_nodelets 
          cmd_vel cmd_vel_out"/>
      <param name="cmd_vel/abtr_priority" value="15"/>
      <node name="do_gesture_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor check_person_nodelets 
          do_gesture do_gesture_out"/>
      <param name="do_gesture/abtr_priority" value="15"/>
    </group>

    <group ns="disengage_interaction">
      <node name="disengage_interaction_nodelets" pkg="nodelet" type="nodelet"
        args="manager"
      />

      <node name="disengage_interaction" pkg="nodelet" type="nodelet"
        args="load hri_behaviors/DisengageInteraction disengage_interaction_nodelets">
        <remap from="pose_to_disengage" to="pose"/>
      </node>

      <node name="pose_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor disengage_interaction_nodelets 
          pose_to_disengage pose"/>

      <node name="pan_angle_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor disengage_interaction_nodelets 
          pan_angle pan_angle_out"/>
      <param name="pan_angle/abtr_priority" value="40"/>
      <node name="cmd_vel_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor disengage_interaction_nodelets 
          cmd_vel cmd_vel_out"/>
      <param name="cmd_vel/abtr_priority" value="40"/>
      <node name="do_gesture_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor disengage_interaction_nodelets 
          do_gesture do_gesture_out"/>
      <param name="do_gesture/abtr_priority" value="40"/>
      <node name="voice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor disengage_interaction_nodelets 
          voice voice_out"/>
      <param name="voice/abtr_priority" value="40"/>
    </group>

    <group ns="engage_interaction">
      <node name="engage_interaction_nodelets" pkg="nodelet" type="nodelet"
        args="manager"
      />

      <node name="engage_interaction" pkg="nodelet" type="nodelet"
        args="load hri_behaviors/EngageInteraction engage_interaction_nodelets">
        <remap from="pose_to_engage" to="pose"/>
      </node>

      <node name="pose_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor engage_interaction_nodelets 
          /sensors/kinect/user_face_pose pose"/>

      <node name="do_gesture_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor engage_interaction_nodelets 
          do_gesture do_gesture_out"/>
      <param name="do_gesture/abtr_priority" value="10"/>
      <node name="voice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor engage_interaction_nodelets 
          voice voice_out"/>
      <param name="voice/abtr_priority" value="10"/>
      <node name="emotion_choice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor engage_interaction_nodelets 
          emotion_choice emotion_choice_out"/>
      <param name="emotion_choice/abtr_priority" value="10"/>
      <node name="pan_angle_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor engage_interaction_nodelets 
          pan_angle pan_angle_out"/>
      <param name="pan_angle/abtr_priority" value="10"/>
    </group>

    <group ns="keep_interacting">
      <node name="keep_interacting_nodelets" pkg="nodelet" type="nodelet"
        args="manager"
      />

      <node name="keep_interacting" pkg="nodelet" type="nodelet"
        args="load hri_behaviors/KeepInteracting keep_interacting_nodelets">
        <remap from="face_pose" to="pose"/>
      </node>

      <node name="pose_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor keep_interacting_nodelets 
          /sensors/kinect/user_face_pose pose"/>

      <node name="pan_angle_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor keep_interacting_nodelets 
          pan_angle pan_angle_out"/>
      <param name="pan_angle/abtr_priority" value="20"/>
      <node name="tilt_angle_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor keep_interacting_nodelets 
          tilt_angle tilt_angle_out"/>
      <param name="tilt_angle/abtr_priority" value="20"/>
      <node name="cmd_vel_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor keep_interacting_nodelets 
          cmd_vel cmd_vel_out"/>
      <param name="cmd_vel/abtr_priority" value="20"/>
      <node name="do_gesture_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor keep_interacting_nodelets 
          do_gesture do_gesture_out"/>
      <param name="do_gesture/abtr_priority" value="20"/>
      <node name="voice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor keep_interacting_nodelets 
          voice voice_out"/>
      <param name="voice/abtr_priority" value="20"/>
      <node name="emotion_choice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor keep_interacting_nodelets 
          emotion_choice emotion_choice_out"/>
      <param name="emotion_choice/abtr_priority" value="20"/>
    </group>

    <group ns="point_at">
      <node name="point_at_nodelets" pkg="nodelet" type="nodelet"
        args="manager"
      />

      <node name="point_at" pkg="nodelet" type="nodelet"
        args="load hri_behaviors/PointAt point_at_nodelets">
        <remap from="pose_to_point" to="pose"/>
      </node>

      <node name="pose_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor point_at_nodelets 
          pose_to_point pose"/>

      <node name="pan_angle_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor point_at_nodelets 
          pan_angle pan_angle_out"/>
      <param name="pan_angle/abtr_priority" value="30"/>
      <node name="cmd_vel_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor point_at_nodelets 
          cmd_vel cmd_vel_out"/>
      <param name="cmd_vel/abtr_priority" value="30"/>
      <node name="do_gesture_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor point_at_nodelets 
          do_gesture do_gesture_out"/>
      <param name="do_gesture/abtr_priority" value="30"/>
      <node name="voice_filter" pkg="nodelet" type="nodelet"
        args="load topic_filters/GenericDivisor point_at_nodelets 
          voice voice_out"/>
      <param name="voice/abtr_priority" value="30"/>
    </group>

    <!-- Registration -->
    <node name="register_emotion" pkg="abtr_priority" type="register"
      args="/face/emotion_choice 
        /attract/emotion_choice 
        /engage_interaction/emotion_choice
        /keep_interacting/emotion_choice" />
    <node name="register_gesture" pkg="abtr_priority" type="register"
      args="/jn0/arm_controller/gesture_choice 
        /attract/do_gesture 
        /check_person/do_gesture 
        /disengage_interaction/do_gesture 
        /engage_interaction/do_gesture 
        /keep_interacting/do_gesture 
        /point_at/do_gesture" /> 
    <node name="register_voice" pkg="abtr_priority" type="register"
      args="/voice/say 
        /attract/voice 
        /disengage_interaction/voice 
        /engage_interaction/voice 
        /keep_interacting/voice 
        /point_at/voice" />
    <node name="register_pan_abs" pkg="abtr_priority" type="register"
      args="/head_ptu/pan_absolute_angle /disengage_interaction/pan_angle"/>
    <node name="register_pan_rel" pkg="abtr_priority" type="register"
      args="/head_ptu/pan_relative_angle 
        /attract/pan_angle 
        /engage_interaction/pan_angle 
        /keep_interacting/pan_angle 
        /point_at/pan_angle" />
    <node name="register_tilt_rel" pkg="abtr_priority" type="register"
      args="/head_ptu/tilt_relative_angle /keep_interacting/tilt_angle"/>

</launch>
